### PR TITLE
Notes for class on Apr06 + Apr13: Programming Models & Serverless

### PR DESCRIPTION
**Pull request by ruparelmetarya**

### CODE REVIEW

File: apr06/msr9732.md [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+   it is inefficient for sharded queries that require shuffles in the data-flow.
+2. It lacks support for some SQL keywords.
+3. Noria’s current design primarily targets relational operators, rather than the iterative or graph computations that 
+   are the focus of other data-flow systems, and processes structured records in tabular form.
+4. The current implementation only shards by hash partitioning on a single column, and resharding requires sending 
+   updates through a single instance, which limits scalability.
+5. The current implementation does not currently support range indices or multi-column joins.
+6. The current implementation re-computes data-flow state on failure.
+
+## Open Questions and Future Work
```
Comment (Ramneek99) [inline apr06/msr9732.md]: One interesting open question would be to do the performance comparison between basic indexing vs Naiad?

File: apr06/msr9732.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    (to all workers, including itself) progress updates. This protocol has an important safety property: no local 
+    frontier ever moves ahead of the global frontier, taken across all outstanding events in the system. A naive 
+    implementation of the protocol would broadcast every progress update and result in impractical communication demands
+    and thus the paper addresses a couple of optimizations on top of it.
+    ![naiad-dpg](msr9732-figs/naiad-dpg.png)
+
+## Trade-Offs
+One of the weakest part of the paper is Fault tolerance and availability. Naiad uses a Checkpoint and Restore mechanism 
+in case of any failure (for the detailed implementation, please refer to the paper). 
+In order to maintain the touted low latency, Naiad does not offer a fast parallel recovery, unlike Spark's DStream.
```
Comment (neeraj71) [inline apr06/msr9732.md]: Would you recommend differential dataflow for more nested and complex computations?

File: apr06/msr9732.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+2. Memory Management Functions - Gives library access to control the amount of memory Weld consumes. This is useful
+   in engines that already perform their own memory management, such as Spark SQL: these engines can call Weld-optimized
+   code on batches of data and spill to disk when they grow past a certain size.
+   
+In nut-shell, the life of a Weld program would look something like:
+
+![weld-lifetime](msr9732-figs/weld-lifetime.png)   
+
+## Trade-Offs
```
Comment (neeraj71) [inline apr06/msr9732.md]: How does Weld perform against split annotations which splits and parallelize data?

File: apr06/msr9732.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+6. Some operators (e.g. projection, filter) are stateless, while some (e.g. count, min/max) are stateful to avoid
+redundant re-computation.
+   
+![noria-view](msr9732-figs/noria-view.png)
+
+An efficient Noria design faces two key challenges:
+1. It must limit the size of its state and views, and 
+2. Changes to the Noria program must adapt the dataflow without downtime in serving clients.
+
+### Partially-stateful data-flow
```
Comment (neeraj71) [inline apr06/msr9732.md]: How does Noria maintain consistency if there are concurrent updates and upquery for a key?

File: apr06/msr9732.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+   presence of new queries and schema changes
+   
+### Dataflow Design
+Noria’s data-flow is a directed acyclic graph of relational operators such as aggregations, joins, and filters. Here
+are the key take-aways from their DataFlow design:
+
+1. Roots of the DAG are base tables
+2. External views are at the leaves.
+3. Internal views are represented by relational operators.
+4. Updates are first applied to the base table and then propagate through the data-flow graph as deltas.
```
Comment (neeraj71) [inline apr06/msr9732.md]: How does it handle if there is an update on missing state for a key? If the value of a key is updated in table "X" and that key is missing in the table "Y" and both "X" and "Y" are joined?

File: apr06/msr9732.md (hunk 2 above)
Comment (ruparelmetarya) [reply apr06/msr9732.md]: Do you mind explaining?

File: apr06/msr9732.md (hunk 4 above)
Comment (ruparelmetarya) [reply apr06/msr9732.md]: Noria is only **eventually consistent** if you've read the paper.

File: apr06/msr9732.md (hunk 3 above)
Comment (ruparelmetarya) [reply apr06/msr9732.md]: What do you mean by that? Weld is a common runtime for  powerful cross-library optimizations. It deals with things like encapsulating multiple loops of same data into one. Like adding +5 to data and finding a mean (just a naive example).

File: apr06/msr9732.md (hunk 5 above)
Comment (ruparelmetarya) [reply apr06/msr9732.md]: I think you should read the `4 Partially-stateful data-flow` part of the paper again. Specifically maybe `4.4 Eviction and recursive upqueries`. They have explained it with example.

**Comment by ruparelmetarya** [comment]
As per my understanding, notes are targeted to give a gist about what the paper holds and explain the techniques and workflows in brevity. If there are conceptual questions related to the papers, I believe office hours or CampuWire would be a better platform for discussions where a broader audience could also participate.

**Comment by apanda** [comment]
> As per my understanding, notes are targeted to give a gist about what the paper holds and explain the techniques and workflows in brevity. If there are conceptual questions related to the papers, I believe office hours or CampuWire would be a better platform for discussions where a broader audience could also participate.

I don't entirely agree: notes should include information that people find useful, and they should evolve in response to comments. In some sense they are meant to complement the papers, not just act as a gist.

**Comment by ruparelmetarya** [comment]
> > As per my understanding, notes are targeted to give a gist about what the paper holds and explain the techniques and workflows in brevity. If there are conceptual questions related to the papers, I believe office hours or CampuWire would be a better platform for discussions where a broader audience could also participate.
> 
> I don't entirely agree: notes should include information that people find useful, and they should evolve in response to comments. In some sense they are meant to complement the papers, not just act as a gist.

Make sense. Got it, Professor!

**Pull request merged**